activerecord/lib/active_record/connection_adapters.rb

```diff
@@ -31,6 +31,48 @@ def resolve(adapter_name) # :nodoc:
         class_name, path_to_adapter = @adapters[adapter_name.to_s]
 
         unless class_name
+          # To provide better error messages for adapters expecting the pre-7.2 adapter registration API, we attempt
+          # to load the adapter file from the old location which was required by convention, and then raise an error
+          # describing how to upgrade the adapter to the new API.
+          legacy_adapter_path = "active_record/connection_adapters/#{adapter_name}_adapter"
+
+          begin
+            require legacy_adapter_path
+            # If we reach here it means we found the found a file that may be the legacy adapter and should raise.
+            message = if ActiveRecord::ConnectionHandling.instance_methods.include?("#{adapter_name}_connection".to_sym)
+              # If we find the connection method then we care certain it is a legacy adapter.
+              <<~MSG.squish
+                Database configuration specifies '#{adapter_name}' adapter but that adapter has not been registered.
+                Rails 7.2 has changed the way Active Record database adapters are loaded. The adapter needs to be
+                updated to register itself rather than being loaded by convention.
+                Ensure that the adapter in the Gemfile is at the latest version. If it is, then the adapter may need to
+                be updated.
+                See:
+                https://edgeapi.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters.html#method-c-register
+                https://github.com/rails/rails/pull/50064
+              MSG
+            else
+              # If we do not find the connection method we are much less certain it is a legacy adapter. Even though the
+              # file exists in the location defined by convenntion, it does not necessarily mean that file is supposed
+              # to define the adapter the legacy way. So raise an error that explains both possibilities.
+              <<~MSG.squish
+                Database configuration specifies nonexistent '#{adapter_name}' adapter.
+                Available adapters are: #{@adapters.keys.sort.join(", ")}.
+                Ensure that the adapter is spelled correctly in config/database.yml and that you've added the necessary
+                adapter gem to your Gemfile if it's not in the list of available adapters.
+                Rails 7.2 has changed the way Active Record database adapters are loaded. Ensure that the adapter in
+                the Gemfile is at the latest version. If it is up to date, the adapter may need to be modified.
+                See:
+                https://edgeapi.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters.html#method-c-register
+                https://github.com/rails/rails/pull/50064
```

Comment:
I think it's also rare that we link to a PR in a deprecation, it's probably TMI for most devs.

This notice is also too long, IMO warnings should get to the point and link for more info, like a guide on how to resolve the warning.